contracts/src/Cluster.sol

```diff
@@ -1,330 +1,555 @@
+    uint8 minOperators;
 }
 
-event MigrationStarted(uint64 id, Keyspace newKeyspace, uint64 newKeyspaceVersion, uint128 clusterVersion);
-event MigrationDataPullCompleted(uint64 id, address operatorAddress, uint128 clusterVersion);
-event MigrationCompleted(uint64 id, address operatorAddress, uint128 clusterVersion);
-event MigrationAborted(uint64 id, uint128 clusterVersion);
-
-event MaintenanceStarted(address addr, uint128 clusterVersion);
-event MaintenanceFinished(uint128 clusterVersion);
-
-event NodeOperatorAdded(uint8 idx, NodeOperator operator, uint128 clusterVersion);
-event NodeOperatorUpdated(NodeOperator operator, uint128 clusterVersion);
-event NodeOperatorRemoved(address operatorAddress, uint128 clusterVersion);
-
-event SettingsUpdated(Settings newSettings, uint128 clusterVersion);
-
-event OwnershipTransferred(address newOwner, uint128 cluserVersion);
-
-contract Cluster {
-    using Bitmask for uint256;
+struct NodeOperator {
+    address addr;
+    bytes data;
+    bool maintenance;
+}
 
-    // TODO: Should we just make all the fields public?
+struct NodeOperatorData {
+    bytes data;
+    bool maintenance;
+}
 
-    address owner;
+struct Keyspace {
+    uint8[] members;
+    uint8 replicationStrategy;
+}
 
-    NodeOperators nodeOperators;
+/// @notice Represents the maintenance state of the cluster
+/// @dev Used to track whether the cluster is in maintenance mode and who initiated it
+/// @param slot The address of the operator or owner who initiated maintenance. address(0) means no maintenance is active
+/// @param isOwnerMaintenance True if maintenance was initiated by the owner, false if by an operator
+struct MaintenanceState {
+    address slot;
+    bool isOwnerMaintenance;
+}
 
-    Keyspace[2] keyspaces;
+struct ClusterView {
+    Settings settings;
+    uint128 version;
     uint64 keyspaceVersion;
+    uint8 operatorCount;
+    address[] operators;
+    NodeOperatorData[] operatorData;
+    Keyspace currentKeyspace;
+    uint64 migrationId;
+    uint8[] pullingOperators;
+    MaintenanceState maintenance;
+}
 
-    Settings settings;
-    Migration migration;
-    Maintenance maintenance;
+contract Cluster is Ownable2StepUpgradeable, UUPSUpgradeable {
+    using LibSort for *;
+
+    /*//////////////////////////////////////////////////////////////////////////
+                            CONSTANTS & ERRORS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    // Maximum number of operators per cluster
+    uint8 constant MAX_OPERATORS = 255;
+
+    error Unauthorized();
+    error MigrationInProgress();
+    error NoMigrationInProgress();
+    error InvalidOperator();
+    error OperatorNotFound();
+    error OperatorNotPulling();
+    error CallerNotOperator();
+    error OperatorExists();
+    error OperatorInKeyspace();
+    error SameKeyspace();
+    error WrongMigrationId();
+    error InvalidOperatorData();
+    error TooManyOperators();
+    error NotInitialized();
```

Comment:
As I mentioned earlier concept of initialization can be removed, as now we have `minOperators`, we can just restrict `minOperators` to be >1

Reply:
I thought that we need it now, because the cluster is not initialized by default (because of the upgradability stuff)

Reply:
> So we dont'allow to go below 1 ever? Remove operator should revert when operatorCount == 1? If so, I can indeed remove this.

So far we never needed it to go below 3. But we can set to 1 or 0, whatever is more convenient.